File: zilli/server/schemas.py
from __future__ import annotations

from pydantic import BaseModel, Field


class RouteRequest(BaseModel):
    request: str = Field(..., min_length=1, max_length=65536)
    industry: str = ""
    force_full_route: bool = False


class RouteResponse(BaseModel):
    final_text: str
    route_type: str
    decision_reason: str
    planner_output: str | None = None
    executor_output: str | None = None
    reviewer_output: str | None = None
    total_duration_ms: float = 0.0
    error: str | None = None


class IndustryRequest(BaseModel):
    request: str = Field(..., min_length=1, max_length=65536)
    force_full_route: bool = False
    sanitize: bool = True
    tenant_id: str = "default"


class IndustryResponse(BaseModel):
    final_text: str
    route_type: str
    decision_reason: str
    industry: str
    tenant_id: str
    total_duration_ms: float = 0.0
    error: str | None = None


class ModelInfo(BaseModel):
    name: str
    backend: str
    model_id: str
    role: str
    alive: bool


class ModelHealth(BaseModel):
    name: str
    model_id: str
    status: str
    base_url: str


class CostStatus(BaseModel):
    remaining_budget: float
    total_calls: int
    calls_this_hour: int
    hourly_quota: float
    emergency_mode: bool


class HealthResponse(BaseModel):
    status: str
    version: str
    models_configured: int
    models_alive: int
    model_health: list[ModelHealth] = []


# ── OpenAI-compatible schemas ─────────────────────────────────────────


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatCompletionRequest(BaseModel):
    model: str = "zilli"
    messages: list[ChatMessage] = Field(..., min_length=1)
    max_tokens: int = 2048
    temperature: float = 0.1
    top_p: float = 0.9
    stop: list[str] | None = None
    n: int = 1
    stream: bool = False


class ChatCompletionChoice(BaseModel):
    index: int
    message: ChatMessage
    finish_reason: str = "stop"


class Usage(BaseModel):
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0


class ChatCompletionResponse(BaseModel):
    id: str
    object: str = "chat.completion"
    created: int
    model: str
    choices: list[ChatCompletionChoice]
    usage: Usage


class OpenAIModel(BaseModel):
    id: str
    object: str = "model"
    created: int
    owned_by: str = "zilli"


class OpenAIModelList(BaseModel):
    object: str = "list"
    data: list[OpenAIModel]

File: zilli-rs/src/server/schemas.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouteRequest {
    pub request: String,
    pub industry: Option<String>,
    pub force_full_route: Option<bool>,
    pub sanitize: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouteResponse {
    pub final_text: String,
    pub route_type: String,
    pub decision: String,
    pub total_duration_ms: u64,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IndustryRequest {
    pub request: String,
    pub industry: String,
    pub tenant_id: Option<String>,
    pub force_full_route: Option<bool>,
    pub sanitize: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionRequest {
    pub model: String,
    pub messages: Vec<ChatMessage>,
    pub max_tokens: Option<i32>,
    pub temperature: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionChoice {
    pub index: i32,
    pub message: ChatMessage,
    pub finish_reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionResponse {
    pub id: String,
    pub object: String,
    pub created: i64,
    pub model: String,
    pub choices: Vec<ChatCompletionChoice>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CostStatus {
    pub remaining_budget: f64,
    pub total_calls: u64,
    pub planner_calls: u64,
    pub executor_calls: u64,
    pub emergency_mode: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthResponse {
    pub status: String,
    pub version: String,
    pub models_healthy: Vec<serde_json::Value>,
}
